File: appbuilder/core/console/rag/rag.py
import json
from appbuilder.core._client import HTTPClient
from appbuilder.core.component import Message, Component
from appbuilder.core.console.base import ConsoleCompletionResponse


class RAG(Component):
    """
    console RAG组件，利用console端RAG应用进行问答
    
    Examples:
    
    .. code-block:: python
    
        import appbuilder
        import os

        os.environ["APPBUILDER_TOKEN"] = '...'
        conversation_id = '...'
        app_id = '...' # 线上知识库ID
        conversation_id = '...' # 会话ID，可选参数，不传默认新建会话
        rag_app = appbuilder.console.RAG(app_id)
        query = "中国的首都在哪里"
        answer = rag_app.run(appbuilder.Message(query)) # 新建会话
        print(answer)
        conversation_id = answer.conversation_id # 获取会话ID，可用于下次会话
        print(conversation_id)
        query = "它有哪些旅游景点"
        answer = rag_app.run(appbuilder.Message(query), conversation_id) # 接上次会话
        print(answer.content)
        print(answer.extra)  # 获取结果来源

    """
    name = "rag"
    integrated_url: str = "/v1/ai_engine/agi_platform/v1/instance/integrated"
    # debug_url: str = "/debug"

    def __init__(self, app_id: str = ""):
        super().__init__()
        self.app_id = app_id
        self._http_client = None

    @property
    def http_client(self):
        if self._http_client is None:
            self._http_client = HTTPClient()
        return self._http_client

    def run(self, query: Message, conversation_id: str = "", stream: bool = False) -> Message:
        """
        RAG问答
        
        Args:
            query: 用户输入的文本
            stream: 是否开启流式模式
            conversation_id: 会话ID，不传表示新建对话
            
        Returns:
            Message: rag答案
        """
        response_mode = "streaming" if stream else "blocking"
        payload = json.dumps({"query": query.content, "app_id": self.app_id,
                              "response_mode": response_mode, "conversation_id": conversation_id})
        headers = self.http_client.auth_header()
        headers["Content-Type"] = "application/json"
        response = self.http_client.session.post(url=self.http_client.service_url(self.integrated_url),
                                                 headers=headers, data=payload, stream=True)
        response = ConsoleCompletionResponse(response, stream)
        return response.to_message()

    def debug(self, query: Message):
        pass

File: go/appbuilder/rag.go
// Copyright (c) 2024 Baidu, Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package appbuilder

import (
	"bufio"
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"time"
)

func NewRAG(appID string, config *SDKConfig) (*RAG, error) {
	if len(appID) == 0 {
		return nil, errors.New("appID is empty")
	}
	if config == nil {
		return nil, errors.New("invalid config")
	}
	client := config.HTTPClient
	if client == nil {
		client = &http.Client{Timeout: 500 * time.Second}
	}
	return &RAG{appID: appID, sdkConfig: config, client: client}, nil
}

type RAG struct {
	appID     string
	sdkConfig *SDKConfig
	client    HTTPClient
}

func (t *RAG) Run(conversationID string, query string, stream bool) (RAGIterator, error) {
	request := http.Request{}
	header := t.sdkConfig.AuthHeader()
	serviceURL, err := t.sdkConfig.ServiceURL("/api/v1/ai_engine/agi_platform/v1/instance/integrated")
	if err != nil {
		return nil, err
	}
	request.URL = serviceURL
	request.Method = "POST"
	header.Set("Content-Type", "application/json")
	request.Header = header
	req := map[string]string{"conversation_id": conversationID,
		"response_mode": "blocking",
		"query":         query,
		"app_id":        t.appID,
	}
	if stream {
		req["response_mode"] = "streaming"
	}
	data, _ := json.Marshal(req)
	request.Body = io.NopCloser(bytes.NewReader(data))
	t.sdkConfig.BuildCurlCommand(&request)
	resp, err := t.client.Do(&request)
	if err != nil {
		return nil, err
	}
	requestID, err := checkHTTPResponse(resp)
	if err != nil {
		return nil, fmt.Errorf("requestID=%s, err=%v", requestID, err)
	}
	r := NewSSEReader(1024*1024, bufio.NewReader(resp.Body))
	if stream {
		return &RAGStreamIterator{requestID: requestID, r: r, body: resp.Body}, nil
	}
	return &RAGOnceIterator{body: resp.Body, requestID: requestID}, nil
}
